File: permutation.py
from typing import List

def permutations(letters: str) -> List[str]:
    # WRITE YOUR BRILLIANT CODE HERE
    # there are multiple ways to do this problem, the tricky part is that we need to verify if the result can be in random order or not
    # if the result needs to be in order then list sliding method will be applied
    # otherwise, we can do the swapping method if the question doesn't ask for the order of the output

    # this method preserves the order
    def backtrack(first, s, curr, res):
        if len(s) == 0:
            res.append("".join(curr[:]))
            return
        for i in range(first, len(s)):
            curr.append(s[i])
            backtrack(first, s[:i] + s[i+1:], curr, res)
            curr.pop() 
    
    result = []
    letterArr = list(letters)
    backtrack(0, letterArr, [], result)
    return result

    # swapping method doesn't preserve the order
    def backtrack_swap_method(first, s, curr, res):
        if len(s) == len(curr):
            res.append("".join(curr[:]))
            return
        for i in range(first, len(s)):
            curr.append(s[i])
            s[first], s[i] = s[i], s[first]
            backtrack(first, s, curr, res)
            s[first], s[i] = s[i], s[first]
            curr.pop() 
    
    result = []
    letterArr = list(letters)
    backtrack(0, letterArr, [], result)
    return result

# solution
def permutations(letters):
    def dfs(path, used, res):
        if len(path) == len(letters):
            res.append(''.join(path))
            return

        for i, letter in enumerate(letters):
            # skip used letters
            if used[i]:
                continue
            # add letter to permutation, mark letter as used
            path.append(letter)
            used[i] = True
            dfs(path, used, res)
            # remove letter from permutation, mark letter as unused
            path.pop()
            used[i] = False

    res = []
    dfs([], [False] * len(letters), res)
    return res

print(permutations("abc"))
print(permutations("ab"))

File: permutation.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
 var permute = function(nums) {
    function backtrack(nums, start, curr, res) {
        if (curr.length === nums.length) {
            res.push([...curr]);
            return
        }
        for (let i = start; i < nums.length; ++i) {
            if (curr.includes(nums[i])) {
                continue;
            }
            curr.push(nums[i]);
            backtrack(nums, start, curr, res);
            curr.pop();
        }
        
    }
    let res = []
    backtrack(nums, 0, [], res);
    return res;
};

//algomonster - my solution
function permutations(letters) {
    // WRITE YOUR BRILLIANT CODE HERE
    function backtrack(letters, curr, res) {
        if (curr.length === letters.length) {
            res.push([...curr].join(""));
            return;
        }
        for (let i = 0; i < letters.length; ++i) {
            if (curr.includes(letters[i])) continue;
            curr.push(letters[i]);
            backtrack(letters, curr, res);
            curr.pop();
        }
        
    }
    const res = [];
    backtrack(letters, [], res);
    return res;
}
//algomonster - solution
function permutations(letters) {
    function dfs(path, used, res) {
        if (path.length == letters.length) {
            return res.push(path.join(''));
        }
        for (let i = 0; i < letters.length; i++) {
            // skip used letters
            if (used[i]) continue;
            // add letter to permutation, mark letter as used
            path.push(letters[i]);
            used[i] = true;
            dfs(path, used, res);
            // remove letter from permutation, mark letter as unused
            path.pop();
            used[i] = false;
        }
    }
    const res = [];
    letters = [...letters];
    dfs([], new Array(letters.length).fill(false), res);
    return res;
}